// @augmentcode/auggie-sdk/dist/context/internal/api-client.js
/**
 * API client for Context operations
 * Handles both indexing endpoints and LLM chat endpoint
 */
import { v4 as uuidv4 } from "uuid";
import { getSDKVersion } from "../../version";
import { retryChat, retryWithBackoff } from "./retry-utils";
/**
 * Error thrown when an API request fails
 * Includes the HTTP status code to enable retry logic for transient failures
 */
export class APIError extends Error {
    constructor(status, statusText, message) {
        super(message);
        this.name = "APIError";
        this.status = status;
        this.statusText = statusText;
    }
}
/**
 * Get user agent string for Context API requests
 */
export function getUserAgent() {
    return `augment.sdk.context/${getSDKVersion()} (typescript)`;
}
/**
 * API client for Context operations
 */
export class ContextAPIClient {
    constructor(options) {
        this.apiKey = options.apiKey;
        this.apiUrl = options.apiUrl;
        this.sessionId = uuidv4();
        this.debug = options.debug ?? false;
    }
    log(message) {
        if (this.debug) {
            console.log(`[ContextAPI] ${message}`);
        }
    }
    createRequestId() {
        return uuidv4();
    }
    async callApi(endpoint, payload, requestId) {
        // Remove trailing slash from apiUrl if present
        const baseUrl = this.apiUrl.endsWith("/")
            ? this.apiUrl.slice(0, -1)
            : this.apiUrl;
        const url = `${baseUrl}/${endpoint}`;
        this.log(`POST ${url}`);
        this.log(`Request ID: ${requestId}`);
        this.log(`Request: ${JSON.stringify(payload, null, 2)}`);
        const response = await fetch(url, {
            method: "POST",
            headers: {
                "Content-Type": "application/json",
                Authorization: `Bearer ${this.apiKey}`,
                "X-Request-Session-Id": this.sessionId,
                "X-Request-Id": requestId,
                "User-Agent": getUserAgent(),
            },
            body: JSON.stringify(payload),
        });
        if (!response.ok) {
            const errorText = await response.text();
            throw new APIError(response.status, response.statusText, `API request failed: ${response.status} ${response.statusText} - ${errorText}`);
        }
        const json = (await response.json());
        this.log(`Response: ${JSON.stringify(json, null, 2)}`);
        return json;
    }
    /**
     * Call API with automatic retry logic and stable request ID across retries
     */
    async callApiWithRetry(endpoint, payload) {
        const requestId = this.createRequestId();
        return await retryWithBackoff(() => this.callApi(endpoint, payload, requestId), this.debug);
    }
    async findMissing(blobNames) {
        const result = await this.callApiWithRetry("find-missing", {
            mem_object_names: blobNames,
        });
        return {
            unknownBlobNames: result.unknown_memory_names || [],
            nonindexedBlobNames: result.nonindexed_blob_names || [],
        };
    }
    async batchUpload(blobs) {
        const result = await this.callApiWithRetry("batch-upload", {
            blobs: blobs.map((blob) => ({
                blob_name: blob.blobName,
                path: blob.pathName,
                content: blob.text,
            })),
        });
        return {
            blobNames: result.blob_names || [],
        };
    }
    async checkpointBlobs(blobs) {
        const result = await this.callApiWithRetry("checkpoint-blobs", {
            blobs: {
                checkpoint_id: blobs.checkpointId ?? null,
                added_blobs: blobs.addedBlobs,
                deleted_blobs: blobs.deletedBlobs,
            },
        });
        return {
            newCheckpointId: result.new_checkpoint_id,
        };
    }
    async agentCodebaseRetrieval(query, blobs, maxOutputLength) {
        const requestBody = {
            information_request: query,
            blobs: {
                checkpoint_id: blobs.checkpointId ?? null,
                added_blobs: blobs.addedBlobs,
                deleted_blobs: blobs.deletedBlobs,
            },
            dialog: [],
        };
        if (maxOutputLength !== undefined) {
            requestBody.max_output_length = maxOutputLength;
        }
        const result = await this.callApiWithRetry("agents/codebase-retrieval", requestBody);
        return {
            formattedRetrieval: result.formatted_retrieval,
        };
    }
    /**
     * Parse a single JSON line from the stream and extract text if present
     */
    parseStreamLine(line) {
        const trimmed = line.trim();
        if (!trimmed) {
            return null;
        }
        try {
            const parsed = JSON.parse(trimmed);
            if (parsed.text) {
                return parsed.text;
            }
        }
        catch {
            // Skip invalid JSON lines
            this.log(`Failed to parse stream line: ${trimmed}`);
        }
        return null;
    }
    /**
     * Parse streaming response and accumulate text chunks
     */
    async parseSSEStream(body) {
        const reader = body.getReader();
        const decoder = new TextDecoder();
        let accumulatedText = "";
        try {
            while (true) {
                const { done, value } = await reader.read();
                if (done)
                    break;
                const chunk = decoder.decode(value, { stream: true });
                const lines = chunk.split("\n");
                for (const line of lines) {
                    const text = this.parseStreamLine(line);
                    if (text) {
                        accumulatedText += text;
                    }
                }
            }
        }
        finally {
            reader.releaseLock();
        }
        return accumulatedText;
    }
    /**
     * Call the LLM chat streaming API with a formatted prompt
     *
     * @param prompt - The formatted prompt to send to the LLM
     * @returns The LLM's response text
     */
    async chat(prompt) {
        const requestId = this.createRequestId();
        return await retryChat(async () => {
            // Remove trailing slash from apiUrl if present
            const baseUrl = this.apiUrl.endsWith("/")
                ? this.apiUrl.slice(0, -1)
                : this.apiUrl;
            const url = `${baseUrl}/chat-stream`;
            this.log(`POST ${url}`);
            this.log(`Request ID: ${requestId}`);
            // Use nodes array format for chat-stream (newer API format)
            // This avoids potential bugs from using the deprecated message field
            // Note: type is an integer enum where 0 = TEXT node
            // The structure matches ChatRequestNode from the backend
            const payload = {
                nodes: [
                    {
                        id: 0,
                        type: 0, // ChatRequestNodeType.TEXT = 0
                        text_node: {
                            content: prompt,
                        },
                    },
                ],
                chat_history: [],
                conversation_id: this.sessionId,
            };
            this.log(`Request: ${JSON.stringify(payload, null, 2)}`);
            const response = await fetch(url, {
                method: "POST",
                headers: {
                    "Content-Type": "application/json",
                    Authorization: `Bearer ${this.apiKey}`,
                    "X-Request-Session-Id": this.sessionId,
                    "X-Request-Id": requestId,
                    "conversation-id": this.sessionId,
                    "X-Mode": "sdk",
                    "User-Agent": getUserAgent(),
                },
                body: JSON.stringify(payload),
            });
            if (!response.ok) {
                const errorText = await response.text();
                throw new APIError(response.status, response.statusText, `API request failed: ${response.status} ${response.statusText} - ${errorText}`);
            }
            // Handle streaming response
            if (!response.body) {
                throw new Error("Response body is null");
            }
            const accumulatedText = await this.parseSSEStream(response.body);
            this.log(`Response: ${accumulatedText}`);
            return accumulatedText;
        }, this.debug);
    }
}
