// @augmentcode/auggie-sdk/dist/context/internal/blob-name-calculator.js
import { createHash } from "node:crypto";
export const blobNamingVersion = 2023102300;
export class BlobTooLargeError extends Error {
    constructor(maxBlobSize) {
        super(`content exceeds maximum size of ${maxBlobSize}`);
    }
}
export class BlobNameCalculator {
    constructor(maxBlobSize) {
        this._textEncoder = new TextEncoder();
        this.maxBlobSize = maxBlobSize;
    }
    _hash(path, contents) {
        const hash = createHash("sha256");
        hash.update(path);
        hash.update(contents);
        return hash.digest("hex");
    }
    calculateOrThrow(path, contents, checkFileSize = true) {
        let contentsBytes;
        if (typeof contents === "string") {
            contentsBytes = this._textEncoder.encode(contents);
        }
        else {
            contentsBytes = contents;
        }
        if (checkFileSize && contentsBytes.length > this.maxBlobSize) {
            throw new BlobTooLargeError(this.maxBlobSize);
        }
        return this._hash(path, contentsBytes);
    }
    calculate(path, contents) {
        try {
            return this.calculateOrThrow(path, contents, true);
        }
        catch {
            return;
        }
    }
    calculateNoThrow(path, contents) {
        return this.calculateOrThrow(path, contents, false);
    }
}
