// @augmentcode/auggie-sdk/dist/context/internal/retry-utils.js
/**
 * Retry utilities with exponential backoff
 * Based on clients/sidecar/libs/src/utils/promise-utils.ts
 */
const defaultBackoffParams = {
    initialMS: 100,
    mult: 2,
    maxMS: 30000,
};
/**
 * Delay for the specified number of milliseconds
 */
function delayMs(ms) {
    if (ms === 0) {
        return Promise.resolve();
    }
    return new Promise((resolve) => {
        setTimeout(resolve, ms);
    });
}
/**
 * Check if an error is retriable based on its properties
 * Retriable errors include network errors and server errors (5xx)
 */
function isRetriableError(e) {
    // Check for HTTP status codes
    if (e && typeof e === "object" && "status" in e) {
        const status = e.status;
        // Match the default VSCode client retry logic:
        // - 499 (cancelled)
        // - 503 (unavailable)
        // - 5xx (unavailable)
        //
        // Note: We do NOT retry 429 or 504 by default
        // (only chat streams retry those)
        return status === 499 || status === 503 || (status >= 500 && status < 600);
    }
    return false;
}
/**
 * Check if an error is retriable for chat streaming endpoints
 * Chat streams retry additional status codes beyond the default retry logic
 */
function isChatRetriableError(e) {
    // Check for HTTP status codes
    if (e && typeof e === "object" && "status" in e) {
        const status = e.status;
        // Chat streams retry:
        // - 429 (rate limit)
        // - 499 (cancelled)
        // - 503 (unavailable)
        // - 529 (overloaded)
        // - 5xx (server errors)
        return (status === 429 ||
            status === 499 ||
            status === 503 ||
            status === 529 ||
            (status >= 500 && status < 600));
    }
    return false;
}
/**
 * Retry a function with exponential backoff
 *
 * @param fn The function to retry
 * @param debug Whether to log debug messages
 * @param backoffParams Backoff parameters
 * @returns Promise that resolves to the function's return value
 */
export async function retryWithBackoff(fn, debug = false, backoffParams = defaultBackoffParams) {
    const canRetryFn = backoffParams.canRetry ?? isRetriableError;
    let backoffMs = 0;
    const startTime = Date.now();
    for (let tries = 0;; tries++) {
        try {
            const ret = await fn();
            if (tries > 0 && debug) {
                console.log(`[RetryUtils] Operation succeeded after ${tries} transient failures`);
            }
            return ret;
        }
        catch (e) {
            const currTryCount = tries + 1;
            // Check if we have exceeded the max number of retries
            if (backoffParams.maxTries !== undefined &&
                currTryCount >= backoffParams.maxTries) {
                throw e;
            }
            // Check if we should retry this error
            if (!canRetryFn(e)) {
                throw e;
            }
            // Calculate backoff delay
            backoffMs =
                backoffMs === 0
                    ? backoffParams.initialMS
                    : Math.min(backoffMs * backoffParams.mult, backoffParams.maxMS);
            if (debug) {
                console.log(`[RetryUtils] Operation failed with error ${e}, retrying in ${backoffMs} ms; retries = ${tries}`);
            }
            // Check if the backoff delay will exceed total time
            if (backoffParams.maxTotalMs !== undefined &&
                Date.now() - startTime + backoffMs > backoffParams.maxTotalMs) {
                throw e;
            }
            await delayMs(backoffMs);
        }
    }
}
/**
 * Retry a chat function with exponential backoff
 * Uses chat-specific retry logic that includes 429 and 529 status codes
 *
 * @param fn The function to retry
 * @param debug Whether to log debug messages
 * @param backoffParams Backoff parameters (defaults to chat-optimized settings)
 * @returns Promise that resolves to the function's return value
 */
export function retryChat(fn, debug = false, backoffParams = defaultBackoffParams) {
    return retryWithBackoff(fn, debug, {
        ...backoffParams,
        canRetry: isChatRetriableError,
    });
}
