// @augmentcode/auggie-sdk/dist/context/internal/__mocks__/api-client.js
/**
 * Mock implementation of ContextAPIClient for testing
 */
import { ContextAPIClient, } from "../api-client";
export class MockContextAPIClient extends ContextAPIClient {
    constructor(config = {}) {
        // Call parent constructor with dummy credentials
        super({
            apiKey: "mock-api-key",
            apiUrl: "https://mock.api.augmentcode.com",
            debug: false,
        });
        // Track method calls for assertions
        this.findMissingCalls = [];
        this.batchUploadCalls = [];
        this.checkpointBlobsCalls = [];
        this.agentCodebaseRetrievalCalls = [];
        this.chatCalls = [];
        this.knownBlobs = config.knownBlobs || new Set();
        this.indexedBlobs = config.indexedBlobs || new Set();
        this.delay = config.delay || 0;
        this.shouldFailFindMissing = config.shouldFailFindMissing ?? false;
        this.shouldFailBatchUpload = config.shouldFailBatchUpload ?? false;
        this.shouldFailCheckpoint = config.shouldFailCheckpoint ?? false;
    }
    createRequestId() {
        return `mock-req-${Date.now()}`;
    }
    async simulateDelay() {
        if (this.delay > 0) {
            await new Promise((resolve) => setTimeout(resolve, this.delay));
        }
    }
    async findMissing(blobNames) {
        this.findMissingCalls.push([...blobNames]);
        if (this.shouldFailFindMissing) {
            throw new Error("Mock findMissing error");
        }
        await this.simulateDelay();
        const unknownBlobNames = blobNames.filter((name) => !this.knownBlobs.has(name));
        const nonindexedBlobNames = blobNames.filter((name) => this.knownBlobs.has(name) && !this.indexedBlobs.has(name));
        return {
            unknownBlobNames,
            nonindexedBlobNames,
        };
    }
    async batchUpload(blobs) {
        this.batchUploadCalls.push([...blobs]);
        if (this.shouldFailBatchUpload) {
            throw new Error("Mock batchUpload error");
        }
        await this.simulateDelay();
        // Add uploaded blobs to known blobs (but not indexed yet)
        const blobNames = blobs.map((b) => b.blobName);
        for (const blobName of blobNames) {
            this.knownBlobs.add(blobName);
        }
        return { blobNames };
    }
    async checkpointBlobs(blobs) {
        this.checkpointBlobsCalls.push({ ...blobs });
        if (this.shouldFailCheckpoint) {
            throw new Error("Mock checkpointBlobs error");
        }
        await this.simulateDelay();
        return {
            newCheckpointId: `checkpoint-${Date.now()}`,
        };
    }
    async agentCodebaseRetrieval(query, blobs, maxOutputLength) {
        this.agentCodebaseRetrievalCalls.push({ query, blobs, maxOutputLength });
        await this.simulateDelay();
        return {
            formattedRetrieval: `Mock search results for: ${query}`,
        };
    }
    async chat(prompt) {
        this.chatCalls.push(prompt);
        await this.simulateDelay();
        return `Mock LLM response to: ${prompt}`;
    }
    // Helper methods for test setup
    /** Mark blobs as indexed (simulates backend indexing completion) */
    markAsIndexed(blobNames) {
        for (const blobName of blobNames) {
            this.indexedBlobs.add(blobName);
        }
    }
    /** Reset all call tracking */
    resetCalls() {
        this.findMissingCalls = [];
        this.batchUploadCalls = [];
        this.checkpointBlobsCalls = [];
        this.agentCodebaseRetrievalCalls = [];
        this.chatCalls = [];
    }
}
