// @augmentcode/auggie-sdk/dist/auggie/sdk-mcp-server.js
import { createServer as createHTTPServer, } from "node:http";
import { createServer } from "node:net";
import { MCPServer } from "@mastra/mcp";
/**
 * Find a free port by creating a temporary server on port 0
 * and getting the assigned port number
 */
function findFreePort() {
    return new Promise((resolve, reject) => {
        const server = createServer();
        server.unref();
        server.on("error", reject);
        server.listen(0, () => {
            const address = server.address();
            if (address && typeof address !== "string") {
                const port = address.port;
                server.close(() => {
                    resolve(port);
                });
            }
            else {
                reject(new Error("Failed to get port from server address"));
            }
        });
    });
}
export class AuggieMCPServer {
    constructor(tools, PORT) {
        this.httpServer = null;
        this.serverPATH = "/mcp";
        this.activeSockets = new Set();
        // Wrap AI SDK tools to match Mastra's expected signature
        // Mastra wraps the actual tool parameters in a 'context' property
        // AI SDK tools expect: execute({ param1, param2 })
        // Mastra passes: execute({ context: { param1, param2 }, runtimeContext, ... }, mcpContext)
        const wrappedTools = {};
        for (const [name, tool] of Object.entries(tools)) {
            wrappedTools[name] = {
                ...tool,
                execute: async (args, mcpContext) => {
                    // Extract the actual tool parameters from the 'context' property
                    const toolParams = args?.context || args;
                    // Call the original AI SDK tool's execute with the unwrapped parameters
                    // AI SDK tools may expect (params, options) signature
                    if (tool.execute) {
                        return await tool.execute(toolParams, mcpContext);
                    }
                    throw new Error(`Tool ${name} does not have an execute function`);
                },
            };
        }
        this.server = new MCPServer({
            name: "auggie-mcp",
            version: "1.0.0",
            tools: wrappedTools,
        });
        this.PORT = PORT;
    }
    async start() {
        // If PORT is undefined, find a free port
        if (this.PORT === undefined) {
            this.PORT = await findFreePort();
        }
        const httpServer = createHTTPServer(async (req, res) => {
            await this.server.startHTTP({
                url: new URL(req.url || "", `http://localhost:${this.PORT}`),
                httpPath: this.serverPATH,
                req,
                res,
                options: {
                    sessionIdGenerator: undefined,
                },
            });
        });
        // Store the server instance
        this.httpServer = httpServer;
        // Track active connections for faster shutdown
        httpServer.on("connection", (socket) => {
            this.activeSockets.add(socket);
            socket.on("close", () => {
                this.activeSockets.delete(socket);
            });
        });
        // Wait for the server to start listening before returning
        await new Promise((resolve, reject) => {
            httpServer.listen(this.PORT, () => {
                resolve();
            });
            httpServer.on("error", reject);
        });
        const serverURL = new URL(this.serverPATH, `http://localhost:${this.PORT}`);
        return serverURL.toString();
    }
    async stop() {
        // Close the HTTP server first with a timeout
        if (this.httpServer) {
            await Promise.race([
                new Promise((resolve, reject) => {
                    this.httpServer?.close((err) => {
                        if (err)
                            reject(err);
                        else
                            resolve();
                    });
                }),
                // Timeout after 1 second
                new Promise((resolve) => setTimeout(resolve, 1000)),
            ]);
            // Force close by destroying all connections if still open
            if (this.httpServer.listening) {
                this.httpServer.closeAllConnections?.();
            }
            this.httpServer = null;
        }
        // Then close the MCP server with a timeout
        await Promise.race([
            this.server.close(),
            // Timeout after 1 second
            new Promise((resolve) => setTimeout(resolve, 1000)),
        ]);
    }
}
